flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/source/FileSplitMonitoringFunction.java

```diff
@@ -0,0 +1,234 @@
+ */
+public class FileSplitMonitoringFunction<OUT>
+	extends RichSourceFunction<FileInputSplit> {
+
+	private static final long serialVersionUID = 1L;
+
+	private static final Logger LOG = LoggerFactory.getLogger(FileSplitMonitoringFunction.class);
+
+	/**
+	 * Specifies when computation will be triggered.
+	 */
+	public enum WatchType {
+		REPROCESS_WITH_APPENDED		// Reprocesses the whole file when new data is appended.
+	}
+
+	/** The path to monitor. */
+	private final String path;
+
+	/** The default parallelism for the job, as this is going to be the parallelism of the downstream readers. */
+	private final int readerParallelism;
+
+	/** The {@link FileInputFormat} to be read. */
+	private FileInputFormat<OUT> format;
+
+	/** How often to monitor the state of the directory for new data. */
+	private final long interval;
+
+	/** Which new data to process (see {@link WatchType}. */
+	private final WatchType watchType;
+
+	private long globalModificationTime;
+
+	private FilePathFilter pathFilter;
+
+	private volatile boolean isRunning = false;
+
+	private Configuration configuration;
+
+	public FileSplitMonitoringFunction(
+		FileInputFormat<OUT> format, String path, Configuration configuration,
+		WatchType watchType, int readerParallelism, long interval) {
+
+		this(format, path, configuration, FilePathFilter.DefaultFilter.getInstance(), watchType, readerParallelism, interval);
+	}
+
+	public FileSplitMonitoringFunction(
+		FileInputFormat<OUT> format, String path, Configuration configuration,
+		FilePathFilter filter, WatchType watchType, int readerParallelism, long interval) {
+
+		this.format = Preconditions.checkNotNull(format);
+		this.path = Preconditions.checkNotNull(path);
+		this.configuration = Preconditions.checkNotNull(configuration);
+
+		Preconditions.checkArgument(interval >= 100,
+			"The specified monitoring interval is smaller than the minimum allowed one (100 ms).");
+		this.interval = interval;
+
+		this.watchType = watchType;
+
+		this.pathFilter = Preconditions.checkNotNull(filter);
+
+		this.readerParallelism = Math.max(readerParallelism, 1);
+		this.globalModificationTime = Long.MIN_VALUE;
+	}
+
+	@Override
+	@SuppressWarnings("unchecked")
+	public void open(Configuration parameters) throws Exception {
+		super.open(parameters);
+		format.configure(this.configuration);
+	}
+
+	/**
+	 * Creates the input splits for the path to be assigned to the downstream tasks.
+	 * Those are going to read their contents for further processing.
+	 */
+	private FileInputSplit[] createInputSplits() throws JobException {
+		FileInputSplit[] inputSplits;
+		try {
+			inputSplits = format.createInputSplits(readerParallelism);
+		} catch (Throwable t) {
+			throw new JobException("Creating the input splits caused an error: " + t.getMessage(), t);
+		}
+		return inputSplits;
+	}
+
+	@Override
+	public void run(SourceFunction.SourceContext<FileInputSplit> context) throws Exception {
+		if(!isRunning) {
```

Comment:
We should not change this here. I think it can be problematic with race conditions. Just initializing `isRunning` to `true` when declaring the field should do the trick. (This is also how the Kafka Consumer does it and I think as this one as the reference consumer since it's most widely used.)